iot-hub/Samples/device/ImportExportDevicesSample/ProcessImpExpCommands.cs

```diff
@@ -0,0 +1,438 @@
+        /// Can use this for testing -- read the devices and export them, 
+        ///   then open the file and see what it has in it.
+        private static async Task ExportDevices(string URI_to_container)
+        {
+            // Create an instance of the registry manager class.
+            RegistryManager registryManager =
+              RegistryManager.CreateFromConnectionString(iotHubConnectionString);
+
+            // Call an export job on the IoT Hub to retrieve all devices.
+            // This writes them to devices.txt in the container. 
+            // The second parameter indicates whether to export the keys or not.
+            JobProperties exportJob = await
+              registryManager.ExportDevicesAsync(URI_to_container, false);
+
+            // Poll every 5 seconds to see if the job has finished executing.
+            while (true)
+            {
+                exportJob = await registryManager.GetJobAsync(exportJob.JobId);
+                if (exportJob.Status == JobStatus.Completed ||
+                    exportJob.Status == JobStatus.Failed ||
+                    exportJob.Status == JobStatus.Cancelled)
+                {
+                    // Job has finished executing
+                    break;
+                }
+
+                await Task.Delay(TimeSpan.FromSeconds(5));
+            }
+        
+            // Note: could add twin data here if you want to export it.
+        }
+
+        /// <summary>
+        /// Read list of exported devices from the blob.
+        /// Deserialize each entry and add it to exportedDevices.
+        /// This is for testing your export.
+        /// </summary>
+        /// <returns></returns>
+        private static async Task ReadExportedDeviceList()
+        {
+            // Get reference to the location of the blob that contains the device list from the IoT Hub.
+            CloudBlockBlob exportedListBlob = cloudBlobContainer.GetBlockBlobReference(deviceListFile);
+
+            // Instantiate a list of devices to read the blob into.
+            List<ExportImportDevice> exportedDevices = new List<ExportImportDevice>();
+
+            // Read the blob. Deserialize each row into an object and write it to exportedDevices.
+            // You end up with a list of devices that exist for the hub in a list in memory.
+            using (var streamReader = new StreamReader(await exportedListBlob.OpenReadAsync(AccessCondition.GenerateIfExistsCondition(), null, null), Encoding.UTF8))
+            {
+                while (streamReader.Peek() != -1)
+                {
+                    string line = await streamReader.ReadLineAsync();
+                    var device = JsonConvert.DeserializeObject<ExportImportDevice>(line);
+                    exportedDevices.Add(device);
+                }
+            }
+
+            // Print out the list of devices to the console.
+            foreach (ExportImportDevice eid in exportedDevices)
```

Comment:
Should this be moved into the reading loop so you don't need to keep the `List<ExportImportDevice>` in memory?

Reply:
no, it's separate on purpose, to thebig that gets the info and puts it in the generic list is separate from the bit that shows it. It's a debug thing, so I can look at the generic list.